### ISSUE TITLE
Improve type inference of Array Literals

### ISSUE THREAD

**Issue by rjchatfield** [nicklockwood/SwiftFormat#886]
Reading this Twitter thread by Ben Cohen, it might be valuable to change all ivar default values to avoid literals. It's uglier (as he admits) but if it improves compiles this significantly then I think it would be awesome!
 
https://twitter.com/airspeedswift/status/1372912670542757891?s=21

Before:

```swift
public struct Foo {
  var arr: [Int] = [] // inferring type of array literal
  var bar = Bar() // inferring type of `bar`
}
```

Suggestion:

```swift 
public struct Foo {
  var arr: [Int] = [Int]()
  var bar: Bar = Bar()
}
```

**Comment by nicklockwood** [comment]
@rjchatfield with the example I think you may have misread the guidance in Ben's tweet. IIUC, it's saying that:

```swift
var arr: [Int] = []
```

Is preferable to:

```swift
var arr = [Int]()
```

But not that you should add redundancy by writing:

```swift
var arr: [Int] = [Int]()
```

**Comment by nicklockwood** [comment]
@rjchatfield sorry, I didn't mean to imply that this issue was invalid, just that the implementation won't be quite as you suggested. There is still value in implementing a feature to insert an explicit type for literals, e.g.

```swift
let foo = 5
let bar = ["hello"]

->

let foo: Int = 5
let bar: [String] = ["hello"]
```

**Comment by rjchatfield** [comment]
Thanks for reopening. 

I think I read somewhere that this inference was most expensive on ivars of a class or struct, especially public types. (Again, feel free to correct me if I'm wrong 😅)

If this feature does make it in to SwiftFormat, I'd appreciate a feature to only apply it to that "class/struct ivar scope"... and opt-out of adding explicit types to every "local variable scope" in an application.

**Comment by nicklockwood** [comment]
@rjchatfield yep, my understanding is that it's a really good idea for compilation speed to use explicit types for globals and properties (less important for locals), and I plan to add more support for that in the next release.

**Comment by nicklockwood** [comment]
@rjchatfield I forgot to update the issue, but this was implemented in 0.49.0
